### PR TITLE
[SDDM] Touchpad tap to select

### PR DESCRIPTION
**Pull request by kRHYME7**
# MERGE REQUEST

Full discussion at https://github.com/prasanthrangan/hyprdots/issues/1081

For laptop/touchpad users:
+ tap to select an option 

Here's the list added to the ```xorg.conf.d/```

```
Section "InputClass"
Identifier "libinput touchpad catchall"
MatchIsTouchpad "on"
MatchDevicePath "/dev/input/event*"
Driver "libinput"
Option "Tapping" "on"
Option "NaturalScrolling" "on"
Option "MiddleEmulation" "on"
Option "DisableWhileTyping" "on"
```

### CODE REVIEW

**Comment by sandeng1440** [comment]
It looks good. I believe that `libinput` is already a dependency for something else in Hyprdots, so this should do it.

**Comment by prasanthrangan** [comment]
is this only valid for x11?? what if you set it to wayland, will this be necessary??

**Comment by sandeng1440** [comment]
@prasanthrangan 
Yes it's only valid for `X11` For `wayland` there's a different way to solve it. I haven't done enough reading on it to know exactly how to solve for `wayland`.

**Comment by kRHYME7** [comment]
We can.( Scanned the wiki a little bit )

@sandeng1440  

Can you please try  remove  the file we added in ``` xorg.conf.d/ ```  (Not needed but just to be accurate)

Then add ``` /etc/sddm.conf.d/10-wayland.conf ``` then append this 

```
[General]
DisplayServer=wayland-user
GreeterEnvironment=QT_WAYLAND_SHELL_INTEGRATION=layer-shell

[Wayland]
CompositorCommand=Hyprland

```

This work on my end.

**Comment by kRHYME7** [comment]
Trying wayland, the touchpad tap to select is not enable by default. Can't find anything to enable it

```
[General]
DisplayServer=wayland-user
GreeterEnvironment=QT_WAYLAND_SHELL_INTEGRATION=layer-shell

[Wayland]
CompositorCommand=Hyprland

```

**Comment by sandeng1440** [comment]
@kRHYME7 
There are 2 compositors:
1. **Hyprland:** It does not run until sddm tells it to. Therefore it cannot be the compositor managing sddm. 
2. **Weston:** I believe is the compositor that is running before sddm and manages sddm.

I found this out when I clicked on the WM selector on sddm.

**Comment by kRHYME7** [comment]
> By default DisplayServer=wayland attempts to use [weston](https://wiki.archlinux.org/title/Weston), even if it's not installed, to use a different Wayland compositor, enter the program and relevant arguments to the CompositorCommand option as described in [SDDM#KDE / KWin](https://wiki.archlinux.org/title/SDDM#KDE_/_KWin), see also [sddm.conf(5)](https://man.archlinux.org/man/sddm.conf.5) for more.

1. Yeah I think Hyprland soon would have it's own DM so this won't be posible
2.  We should find an option for weston to  have tap-to-click.

**Comment by sandeng1440** [comment]
@kRHYME7 
It looks like KWin will be easy to add. You could try making `KWin` the compositor.

**Comment by sandeng1440** [comment]
# Wayland Solution FOUND
---
I used `weston` as the compositor. It is the default wayland compositor set in sddm. 
All I had to do is add `/etc/xdg/weston/weston.ini` and then populated it with the following contents found at [this arch wiki on weston](https://wiki.archlinux.org/title/Weston)

```ini
[core]
# xwayland support
xwayland=true

[libinput]
enable-tap=true

[input]
touchpad_tap_to_click=true

[shell]
#background-image=/usr/share/backgrounds/gnome/Aqua.jpg
background-type=scale-crop
background-color=0xff000000
#background-color=0xff002244
#panel-color=0x90ff0000
panel-color=0x00ffffff
panel-position=bottom
#clock-format=none
#animation=zoom
#startup-animation=none
close-animation=none
focus-animation=dim-layer
#binding-modifier=ctrl
num-workspaces=6
locking=false
cursor-theme=Adwaita
cursor-size=24

# tablet options
#lockscreen-icon=/usr/share/icons/gnome/256x256/actions/lock.png
#lockscreen=/usr/share/backgrounds/gnome/Garden.jpg
#homescreen=/usr/share/backgrounds/gnome/Blinds.jpg
#animation=fade

# for Laptop displays
[output]
name=LVDS1
mode=preferred
#mode=1680x1050
#transform=rotate-90

#[output]
#name=VGA1
# The following sets the mode with a modeline, you can get modelines for your preffered resolutions using the cvt utility
#mode=173.00 1920 2048 2248 2576 1080 1083 1088 1120 -hsync +vsync
#transform=flipped

#[output]
#name=X1
#mode=1024x768
#transform=flipped-rotate-270

# on screen keyboard input method
#[input-method]
#path=/usr/lib/weston/weston-keyboard

[keyboard]
keymap_rules=evdev
#keymap_layout=us,de
#keymap_variant=colemak,
#keymap_options=grp:shifts_toggle
#keymap_options=caps:ctrl_modifier,shift:both_capslock_cancel
repeat-rate=30
repeat-delay=300

# keymap_options from /usr/share/X11/xkb/rules/base.lst
#numlock-on=true

[terminal]
font=monospace
font-size=18

[launcher]
icon=/usr/share/weston/icon_flower.png
path=/usr/bin/weston-flower

[launcher]
icon=/usr/share/icons/gnome/32x32/apps/utilities-terminal.png
path=/usr/bin/weston-terminal --shell=/usr/bin/bash

#[launcher]
#icon=/usr/share/icons/gnome/32x32/apps/utilities-terminal.png
#path=/usr/bin/gnome-terminal

[launcher]
icon=/usr/share/icons/hicolor/32x32/apps/firefox.png
path=MOZ_ENABLE_WAYLAND=1 /usr/bin/firefox

#[launcher]
#icon=/usr/share/icons/Adwaita/32x32/apps/multimedia-volume-control.png
#path=/usr/bin/st alsamixer -c0
```

Here's what my `/etc/sddm.conf.d/sddm.conf` looks like:
```ini
[core]
# xwayland support
xwayland=true

[libinput]
enable-tap=true

[shell]
#background-image=/usr/share/backgrounds/gnome/Aqua.jpg
background-type=scale-crop
background-color=0xff000000
#background-color=0xff002244
#panel-color=0x90ff0000
panel-color=0x00ffffff
panel-position=bottom
#clock-format=none
#animation=zoom
#startup-animation=none
close-animation=none
focus-animation=dim-layer
#binding-modifier=ctrl
num-workspaces=6
locking=false
cursor-theme=Adwaita
cursor-size=24

# tablet options
#lockscreen-icon=/usr/share/icons/gnome/256x256/actions/lock.png
#lockscreen=/usr/share/backgrounds/gnome/Garden.jpg
#homescreen=/usr/share/backgrounds/gnome/Blinds.jpg
#animation=fade

# for Laptop displays
[output]
name=LVDS1
mode=preferred
#mode=1680x1050
#transform=rotate-90

#[output]
#name=VGA1
# The following sets the mode with a modeline, you can get modelines for your preffered resolutions using the cvt utility
#mode=173.00 1920 2048 2248 2576 1080 1083 1088 1120 -hsync +vsync
#transform=flipped

#[output]
#name=X1
#mode=1024x768
#transform=flipped-rotate-270

# on screen keyboard input method
#[input-method]
#path=/usr/lib/weston/weston-keyboard

[keyboard]
keymap_rules=evdev
#keymap_layout=us,de
#keymap_variant=colemak,
#keymap_options=grp:shifts_toggle
#keymap_options=caps:ctrl_modifier,shift:both_capslock_cancel
repeat-rate=30
repeat-delay=300

# keymap_options from /usr/share/X11/xkb/rules/base.lst
#numlock-on=true

[terminal]
font=monospace
font-size=18

[launcher]
icon=/usr/share/weston/icon_flower.png
path=/usr/bin/weston-flower

[launcher]
icon=/usr/share/icons/gnome/32x32/apps/utilities-terminal.png
path=/usr/bin/weston-terminal --shell=/usr/bin/bash

#[launcher]
#icon=/usr/share/icons/gnome/32x32/apps/utilities-terminal.png
#path=/usr/bin/gnome-terminal

[launcher]
icon=/usr/share/icons/hicolor/32x32/apps/firefox.png
path=MOZ_ENABLE_WAYLAND=1 /usr/bin/firefox

#[launcher]
#icon=/usr/share/icons/Adwaita/32x32/apps/multimedia-volume-control.png
#path=/usr/bin/st alsamixer -c0
```

**Comment by sandeng1440** [comment]
# TLDR
---
Basically, the only change I made is add `/etc/xdg/weston/weston.ini` and then changed the `DisplayServer` option in sddm from `DisplayServer=x11` to `DisplayServer=wayland`

**Comment by kRHYME7** [comment]
Cool! Only problem we might face is if weston is not present by default 
How did you get your weston? i have 2 machines, my other one do not have weston. 

```
paru -Qi weston    
```

Also, should it be fine if we added weston as dependency so that we can composite sddm on wayland?

**Comment by sandeng1440** [comment]
@kRHYME7 
I didn't install weston myself. I think it's a dependency of something in Hyprdots. Lemme setup a VM to see if it's installed by Hyprdots.

**Comment by kRHYME7** [comment]
Try to run ``` pacman -Qi weston     ``` maybe it give you lead to why this is installed

**Comment by sandeng1440** [comment]
@kRHYME7 
That will be sufficient.

**Comment by sandeng1440** [comment]
@kRHYME7 
Here is the result of `yay -Qi weston`

```
Name            : weston
Version         : 13.0.0-1
Description     : Reference implementation of a Wayland compositor
Architecture    : x86_64
URL             : https://wayland.freedesktop.org/
Licenses        : MIT
Groups          : None
Provides        : None
Depends On      : glibc  wayland  libxkbcommon  libinput  libunwind  pixman  libdrm  pam  systemd-libs  cairo  libpng  libjpeg-turbo  libwebp  mesa
                  libegl  libgles  glib2  pango  lcms2  mtdev  libx11  libxcb  dbus  libva  libxcursor  colord  seatd  gstreamer
                  gst-plugins-base-libs
Optional Deps   : xorg-xwayland: support x11 backend [installed]
                  libpipewire: support pipewire backend [installed]
                  freerdp: support rdp backend
                  neatvnc: support vnc backend
Required By     : None
Optional For    : xwayland-run-git
Conflicts With  : None
Replaces        : None
Installed Size  : 5.92 MiB
Packager        : Sébastien Luttringer <seblu@archlinux.org>
Build Date      : Sun 17 Dec 2023 01:57:38 PM EAT
Install Date    : Wed 21 Feb 2024 08:18:01 AM EAT
Install Reason  : Installed as a dependency for another package
Install Script  : No
Validated By    : Signature
```

**Comment by kRHYME7** [comment]
I had a fresh install of hyprdots and weston is not present.

**Comment by prasanthrangan** [comment]
same here, `weston` is not installed by default... but for some reason setting sddm DisplayServer to wayland still works.... which is strange.....

**Comment by kRHYME7** [comment]
>same here, weston is not installed by default... but for some reason setting sddm DisplayServer to wayland still works.... which is strange.....

Maybe on fallback mode?  Or hyprland now can manage sddm layer?

**Comment by kRHYME7** [comment]
Draft this as  sddm actually can recognize hyprland in test-mode
image

**Comment by kRHYME7** [comment]
@sandeng1440  Hi! 

This time we can manage to render sddm on Hyprland 

see comments below
https://github.com/hyprwm/Hyprland/issues/820

The only issue we might get is sddm only renders on a single monitor.  Can be solved if we mirror them with some caveats
image

Will be Adding the config in [this](https://github.com/hyprwm/Hyprland/issues/820) discussion for now, someone might refactor/refine it.

**Comment by sandeng1440** [comment]
@kRHYME7
I've read the comments and I understand that hyprland desktop appears first, followed by sddm. Isn't this behavior a problem?

**Comment by kRHYME7** [comment]
>  I've read the comments and I understand that hyprland desktop appears first, followed by sddm. Isn't this behavior a problem?

Well, not really. The only issue I can think of is we will have 2 Hyprland sessions. 1 is for sddm (after login, sddm will be the parent session) and 1 for the User. Which roughly consumes around 240 mb of memory for sddm. I do not know how much memory weston/cage will consume when running sddm though.

**Comment by kRHYME7** [comment]
@sandeng1440  I'm totally wrong.  Only a single Hyprland instance is active after loging in from sddm using hyprland. 

 Only Issue I got is when I plug in a monitor, sddm will create a new window then now we have 2 windows inside a single monitor instead of this windows should be on the new monitor. Can be solved with scripting and IPC. 
 
 Sorry for being too enthusiastic with using hyprland as compositor for sddm lol.

**Comment by sandeng1440** [comment]
@kRHYME7 no worries. This seems to be a harder issue than I first thought.

**Comment by kRHYME7** [comment]
@sandeng1440  Finally. 

I manage to make a repo for [sddm-hyprland](https://github.com/HyDE-Project/sddm-hyprland) 


Thanks to this [thread ](https://github.com/hyprwm/Hyprland/issues/820#issuecomment-2440060611)
